### PR TITLE
Use detect-port to ensure port is free in tinylicious

### PR DESCRIPTION
**Pull request by elav000**
While running a create-react-app website on its default port 3000, I noticed that tinylicious doesn't complain or throw an error if it is started and things do not work as expected.

I added a check using detect-port to make sure that a port is free before starting the server.

### CODE REVIEW

File: server/tinylicious/src/runner.ts [hunk 1]
```diff
@@ -34,37 +35,42 @@ export class TinyliciousRunner implements utils.IRunner {
...
-            this.contentCollection,
-            new TestClientManager(),
-            new DefaultMetricClient(),
-            winston);
-
-        // Listen on provided port, on all network interfaces.
-        httpServer.listen(this.port);
-        httpServer.on("error", (error) => this.onError(error));
-        httpServer.on("listening", () => this.onListening());
+        this.ensurePortIsFree()
```
Comment (anthony-murphy) [inline server/tinylicious/src/runner.ts:43]: same here. i don't think the deferred is necessesary, and await and try catch is preferred over then/catch, as it keeps the control flow eaiser to understand and modify.

File: server/tinylicious/src/runner.ts [hunk 2]
```diff
@@ -77,11 +83,38 @@ export class TinyliciousRunner implements utils.IRunner {
...
                 this.runningDeferred.reject(error);
-            });
+            },
+        );
 
         return this.runningDeferred.promise;
     }
 
+    private ensurePortIsFree(): Promise<void> {
+        const detectDeferred = new Deferred<void>();
```
[the remark this answers is not in the document]
Comment (elav000) [reply server/tinylicious/src/runner.ts:93]: Great, glad to hear it. I'll remove it. Thanks!

File: server/tinylicious/src/runner.ts [hunk 3]
```diff
@@ -34,21 +35,21 @@ export class TinyliciousRunner implements utils.IRunner {
...
 
-        // Create the HTTP server and attach alfred to it
-        const alfred = app.create(
-            this.config,
-            this.storage,
-            this.mongoManager);
+        try {
+            await this.ensurePortIsFree();
+        } catch (error) {
+            this.runningDeferred.reject(error);
```
Comment (anthony-murphy) [inline server/tinylicious/src/runner.ts:46]: this won't end control flow? should we just let ensurePort throw?

Comment (elav000) [reply server/tinylicious/src/runner.ts:46]: Agreed. Will make it so. In the future we can probably ask whether the individual would like to continue and ask for input (and suggest the free port returned by ensurePortIsFree).

**Review by anthony-murphy — APPROVED** [review]

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**